feat: library home page bare bones (TEMP)

rpenido opened the pull request:
## Description
This PR adds a "Bare Bones" version of the Library Home and updates the SearchResult component to redirect to this new page.

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/0611fdfa-e357-471d-9561-7ea0ee0a696a)

![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/e1bf1807-6a02-47e3-a268-38d210897687)

## More Info
- Closes: https://github.com/openedx/frontend-app-course-authoring/issues/1036
- Depends on: https://github.com/openedx/frontend-app-course-authoring/pull/1050

## Testing Instructions
- Run your local stack on this branch
- Make sure you have the studio.library_authoring_mfe waffle flag **DISABLED**
- Go to the Studio Home and check the "Libraries" tab
- Click on one library with components
- Check if the component count is shown in the last card and the "Components" tab
- Use the keyword search box and check if the component count is filtered accordingly
- Navigate through the Home, Components and Collections tab and check if you can navigate back using the "Back" button from the brownser
- Go back to the Studio Home and open a library WITHOUT components
- You should see a message with a button that will allow the user to add new components (not implemented yet)
- Go back to the Studio Home and open the search modal
- Find a library v2 component and click on it. It should redirect to the new Library Home.    

---
Private ref: [FAL-3753](https://tasks.opencraft.com/browse/FAL-3753)

yusuf-musleh committed:
feat: Add lib v2/legacy tabs in studio home

When lib mode is set to "mixed", both "Libraries" and "Legacy Libraries" tabs are show in the Studio Home. When "Libraries" is clicked, v2 libraries are fetched, when "Legacy Libraries" is clicked, v1 libraries are fetched.

When lib mode is set to "v1 only" or "v2 only", only one tab "Libraries" is show and only the respective libraries are fetched when the tab is clicked.

yusuf-musleh committed:
feat: Add lib v2/legacy tabs in studio home

When lib mode is set to "mixed", both "Libraries" and "Legacy Libraries" tabs are show in the Studio Home. When "Libraries" is clicked, v2 libraries are fetched, when "Legacy Libraries" is clicked, v1 libraries are fetched.

When lib mode is set to "v1 only" or "v2 only", only one tab "Libraries" is show and only the respective libraries are fetched when the tab is clicked.

yusuf-musleh committed:
feat: Add `LIBRARY_MODE` config variable

This is to switch between different library modes.

yusuf-musleh committed:
feat: Add `LIBRARY_MODE` config variable

This is to switch between different library modes.

yusuf-musleh committed:
feat: Add url paths/navigation for each tab

The path updates when selecting tabs, when accessing the url with the path directly it will open its respective tab. Navigating using the browser back/forward buttons is also supported.

yusuf-musleh committed:
feat: Add url paths/navigation for each tab

The path updates when selecting tabs, when accessing the url with the path directly it will open its respective tab. Navigating using the browser back/forward buttons is also supported.

yusuf-musleh committed:
feat: LibraryV2 redirect to lib mfe or placeholder

yusuf-musleh committed:
feat: LibraryV2 redirect to lib mfe or placeholder

yusuf-musleh committed:
feat: Add pagination support for lib v2s

yusuf-musleh committed:
feat: Add pagination support for lib v2s

yusuf-musleh committed:
fix: Redirect to placeholder create lib in v2/mixed disabled mfe

yusuf-musleh committed:
fix: Redirect to placeholder create lib in v2/mixed disabled mfe

yusuf-musleh committed:
temp: This removes TS code to get tests to run

This commit is temporary as the current frontend build system in tests doesnt support TS syntax. That should be fixed soon, and this commit should be removed.

yusuf-musleh committed:
temp: This removes TS code to get tests to run

This commit is temporary as the current frontend build system in tests doesnt support TS syntax. That should be fixed soon, and this commit should be removed.

yusuf-musleh committed:
test: Update existing tests to support changes

yusuf-musleh committed:
test: Update existing tests to support changes

yusuf-musleh committed:
temp: Rename .tsx -> .jsx & .ts -> .js for tests

This is a temporary commit since there are currently no webpack loaders that support tsx files in the test running. This commit should be removed once that is fixed upstream.

yusuf-musleh committed:
temp: Rename .tsx -> .jsx & .ts -> .js for tests

This is a temporary commit since there are currently no webpack loaders that support tsx files in the test running. This commit should be removed once that is fixed upstream.

yusuf-musleh committed:
fix: Fix lint issues

yusuf-musleh committed:
fix: Fix lint issues

rpenido committed:
feat: library home page bare bones

yusuf-musleh committed:
test: Add tests for new functionality

yusuf-musleh committed:
test: Add tests for new functionality

rpenido committed:
fix: update search modal for new library urls

yusuf-musleh commented on src/index.jsx:
```diff
@@ -55,7 +55,7 @@ const App = () => {
         <Route path="/legacy-libraries" element={<StudioHome />} />
-        <Route path="/library/:libraryId" element={<LibraryV2Placeholder />} />
+        <Route path="/library/:libraryId/*" element={<LibraryAuthoringPage />} />
         <Route path="/course/:courseId/*" element={<CourseAuthoringRoutes />} />
```
If you click on the "Create Library" button when the `studio.library_authoring_mfe` flag is **disabled**, it navigates to the `http://apps.local.edly.io:2001/course-authoring/library/create` (which renders the placeholder page as a temporary thing I added in my PR), however that's broken because it assumes `create`  is a libraryKey and the backend errors out. 

Not sure where it should redirect to, maybe a new placeholder "create" page?

rpenido replied:
I created a placeholder for the New Library page (a24b3ba35bc129248a4983d724b680a7b7e83b08)
Also, I didn't find a task to create this form in the course authoring. CC @ChrisChV

ChrisChV replied:
@rpenido Yes, there is no such screen in the mockups. @bradenmacdonald I guess that's included in another epic?

bradenmacdonald replied:
Ah, we missed that. I'll create a story for it. For now, just use the placeholder.

bradenmacdonald replied:
Here is the story for it: https://github.com/openedx/frontend-app-course-authoring/issues/1077

yusuf-musleh commented on src/library-authoring/EmptyStates.jsx:
```diff
@@ -0,0 +1,21 @@
+export const NoComponents = () => (
+  <Stack direction="horizontal" gap={3} className="mt-6 justify-content-center">
+    <div>You have not added any content to this library yet.</div>
+    <Button iconBefore={Add}>Add Content</Button>
```
nit: The text here in both the div and button is missing internationalization.

rpenido replied:
Thank you for catching that! Fixed: 4deaea9a1f3d70aff3d175035f97947c5e996618

yusuf-musleh commented on src/library-authoring/LibraryHome.jsx:
```diff
@@ -0,0 +1,61 @@
+  return (
+    <Stack gap={3}>
+      <Section title="Recently Modified">
+        Recently modified components and collections will be displayed here.
```
nit: Similarly here, internationalize text.

rpenido replied:
Fixed: 4deaea9a1f3d70aff3d175035f97947c5e996618

yusuf-musleh commented on src/search-modal/SearchResult.jsx:
```diff
@@ -149,10 +144,16 @@ const SearchResult = ({ hit }) => {
-        return getLibraryHitUrl(hit, libraryAuthoringMfeUrl);
+    if (contextKey.startsWith('lib:')) {
+      const urlSuffix = getLibraryComponentUrlSuffix(hit);
+      if (libraryAuthoringMfeUrl) {
+        return `${libraryAuthoringMfeUrl}${urlSuffix}`;
       }
```
When clicking on the search result it still redirects to the library authoring mfe when the flag is disabled. I think it missing `&& redirectToLibraryAuthoringMfe` in the check.

rpenido replied:
Fixed: c2bdecfae35bea464a439f3043faddd305c76f01

yusuf-musleh reviewed (changes requested):
@rpenido Great work on this! Overall it works pretty well, just found a things that need to be fixed and some small nits.

yusuf-musleh committed:
refactor: Change /legacy-libraries -> /libraries-v1

rpenido committed:
fix: add i18n messages

rpenido committed:
fix: libraryAuthoring enabled check

rpenido committed:
Merge branch 'yusuf-musleh/lib-v2-tab-studio-home' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: add Create Library placeholder

yusuf-musleh committed:
refactor: Remove hardcoded mfe path

rpenido committed:
refactor: rename .ts files to .js

yusuf-musleh committed:
feat: Add function to construct Lib Auth MFE URL

yusuf-musleh committed:
feat: Make URL /library-v1 when referencing legacy

yusuf-musleh committed:
fix: Add missing part of path

rpenido committed:
fix: type and lint errors

yusuf-musleh committed:
fix: Issue with destinationUrl

yusuf-musleh committed:
test: Add checks for Tab.eventKey in tests

yusuf-musleh committed:
fix: Revert card item url changes to keep simple

rpenido committed:
fix: add tests

rpenido committed:
Merge branch 'yusuf-musleh/lib-v2-tab-studio-home' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: removing unused file

rpenido committed:
fix: add ts-check

rpenido committed:
fix: removing deleted file references

rpenido committed:
chore: trigger CI

yusuf-musleh commented on src/library-authoring/LibraryAuthoringPage.jsx:
```diff
@@ -90,7 +95,6 @@ const LibraryAuthoringPage = () => {
         <SearchField
           value={searchKeywords}
           placeholder={intl.formatMessage(messages.searchPlaceholder)}
-          onSubmit={(value) => setSearchKeywords(value)}
```
The `onSubmit` is needed otherwise if you press 'Enter' it will error out: `onSubmit is not a function`. To avoid performing the search again we can just pass in a no-op function `() => {}`.

yusuf-musleh replied:
I realized this is the same comment as the one I added below.

yusuf-musleh commented on src/library-authoring/LibraryAuthoringPage.test.jsx:
```diff
@@ -0,0 +1,237 @@
+      index_name: 'studio',
+      api_key: 'test-key',
+    });
+    //
```
nit: extra blank comment

rpenido replied:
Fixed: 388c40e2ffd072d8c1fe770f777d8d7bea9bfe3a

yusuf-musleh commented on webpack.dev-tutor.config.js:
Seems like this file slipped thought and got pushed accidentally.

rpenido replied:
Fixed: 388c40e2ffd072d8c1fe770f777d8d7bea9bfe3a

yusuf-musleh commented on src/library-authoring/LibraryAuthoringPage.jsx:
```diff
@@ -0,0 +1,135 @@
+          title={<SubHeaderTitle title={libraryData.title} />}
+          subtitle={intl.formatMessage(messages.headingSubtitle)}
+        />
+        <SearchField
+          value={searchKeywords}
+          placeholder={intl.formatMessage(messages.searchPlaceholder)}
+          onChange={(value) => setSearchKeywords(value)}
+          className="w-50"
+        />
```
The `onSubmit` still needs to be defined here, as if you press "Enter" it causes an error `onSubmit is not a function`. We can pass in a no-op function `() => {}` since the search is already performed `onChange` so it doesn't fire again.

rpenido replied:
Good catch! Fixed: 388c40e2ffd072d8c1fe770f777d8d7bea9bfe3a

yusuf-musleh reviewed (approved):
@rpenido Great work here! Thanks for addressing the issues, things are working well!

Left a few final nits/comments.

rpenido committed:
fix: fixes from review

bradenmacdonald commented on src/search-modal/data/apiHooks.js:
```diff
@@ -45,6 +45,9 @@ export const useContentSearchResults = ({
   blockTypesFilter,
   tagsFilter,
 }) => {
+  blockTypesFilter ??= []; // eslint-disable-line no-param-reassign -- default value for optional parameter
+  tagsFilter ??= []; // eslint-disable-line no-param-reassign -- Default value for optional parameter
+
```
There's a simpler way to write this:

```suggestion
  blockTypesFilter = [],
  tagsFilter = [],
}) => {
```

rpenido replied:
Thank you @bradenmacdonald!
Done: 7d6096e2cea78aa79497305058d019fda31e5a35

rpenido committed:
fix: fix default parameter syntax

rpenido committed:
fix: new library redirect

rpenido committed:
Revert "fix: new library redirect"

This reverts commit c63bc2fbf586a791770b3b63e3262c5c6bd542ec.

rpenido committed:
chore: trigger CI

rpenido committed:
fix: new library redirect

rpenido committed:
fix: remove warning on library home

rpenido committed:
fix: remove unused import

yusuf-musleh committed:
feat: Add lib v2/legacy tabs in studio home (#1050)

This PR adds a new configuration flag that shows/hides tabs in studio home along with some new functionality around to V1 and V2 Libraries.

When the new LIBRARY_MODE flag is set to "mixed" (default in dev) it will show "Libraries" and "Legacy Libraries" tabs that correspond to v1 and v2 tabs respectively.

When the new LIBRARY_MODE flag is set to "v1 only" (default in production) or "v2 only", only one tab "Libraries" is shown and only the respective libraries are fetched when the tab is clicked.

In addition to the above changes, the URL/route now updates when clicking on the tabs, and navigating to it directly would open up that tab as well as a new placeholder page that you will be redirected to when clicking on a v2 library if the library authoring MFE is not enabled.

rpenido committed:
Merge branch 'master' into rpenido/fal-3753-library-home-page-bare-bones

rpenido committed:
fix: merging nits
